File: backend/agents/resumeRewriterAgent.js
const { callGroq } = require('../services/groqService');
const { parseJSON, normalizeRewrittenResume } = require('../utils/responseParser');
const { RAGPipeline } = require('../rag/ragPipeline');

/**
 * AGENT 6: Resume Rewriter Agent
 *
 * Rewrites key resume sections tailored to the target role.
 * Returns: professionalSummary, improvedProjects, improvedSkillsSection, additionalTips
 */
const runResumeRewriterAgent = async (ragPipeline, targetRole) => {
  console.log('\n🤖 Agent 6: Resume Rewriter running...');

  const resumeContext = await ragPipeline.retrieve(
    'professional summary objective projects experience skills achievements',
    8,
    'resume'
  );
  const jdContext = await ragPipeline.retrieve(
    'role expectations keywords preferred qualifications responsibilities',
    4,
    'jd'
  );

  const systemPrompt = `You are a Professional Resume Writer with expertise in crafting ATS-optimized, compelling resumes for top tech companies. You specialize in transforming ordinary resumes into powerful career documents. Use strong action verbs, quantify impact, and include relevant keywords.`;

  const userPrompt = `Rewrite key sections of this resume to be perfectly tailored for a ${targetRole} role.

ORIGINAL RESUME CONTENT:
${RAGPipeline.truncate(resumeContext, 2000)}

JOB REQUIREMENTS:
${RAGPipeline.truncate(jdContext, 1000)}

TARGET ROLE: ${targetRole}

Rewrite rules:
1. Professional Summary: 3-4 impactful sentences with role-specific keywords
2. Projects: Each description should start with an action verb, mention tech stack, and show impact
3. Skills: Organize by category (Frontend, Backend, Tools, Databases, etc.)
4. Additional Tips: Role-specific advice

Return ONLY a JSON object:
{
  "professionalSummary": "Dynamic software engineer with X years of experience...",
  "improvedProjects": [
    "Developed [Project] using [Tech Stack], achieving [impact/result]",
    "Built [Project] with [Tech Stack], reducing [metric] by X%",
    "Designed and implemented [Project] leveraging [Tech Stack] for [outcome]"
  ],
  "improvedSkillsSection": "Frontend: React, Redux, TypeScript | Backend: Node.js, Express | Database: MongoDB, PostgreSQL | Tools: Git, Docker",
  "additionalTips": [
    "Add a GitHub profile link with active repositories",
    "Quantify each achievement with metrics",
    "Include a link to a deployed project"
  ]
}`;

  const raw = await callGroq(systemPrompt, userPrompt, { temperature: 0.4, maxTokens: 3000 });
  const parsed = parseJSON(raw, {});
  return normalizeRewrittenResume(parsed);
};

module.exports = { runResumeRewriterAgent };

File: ai-service/agents/resume_rewriter_agent.py
"""
Agent 6: Resume Rewriter Agent
================================
Rewrites key resume sections tailored to the target role.

Output:
    {
        "professionalSummary": "...",
        "improvedProjects": [...],
        "improvedSkillsSection": "...",
        "additionalTips": [...]
    }
"""

from rag.rag_pipeline import RAGPipeline
from services.groq_service import call_groq
from utils.response_parser import parse_json, normalize_rewritten_resume
from prompts import resume_rewriter_prompt


async def run_resume_rewriter_agent(
    rag: RAGPipeline,
    target_role: str,
) -> dict:
    """
    AGENT 6: Resume Rewriter

    Retrieves the full resume content and rewrites key sections
    (summary, projects, skills) to be more impactful and ATS-optimized
    for the target role.

    Args:
        rag:         Initialized RAGPipeline instance
        target_role: Target job role for tailoring

    Returns:
        Normalized rewritten resume dict with improved sections and tips
    """
    print("\n🤖 Agent 6: Resume Rewriter running...")

    # ─── RAG Retrieval ────────────────────────────────────────
    # Get comprehensive resume content for rewriting
    resume_context = await rag.retrieve(
        "professional summary objective experience projects skills technologies achievements",
        top_k=8,
        source="resume",
    )

    # ─── Prompt Construction ──────────────────────────────────
    system_prompt = resume_rewriter_prompt.get_system_prompt()
    user_prompt = resume_rewriter_prompt.get_user_prompt(
        resume_context=RAGPipeline.truncate(resume_context, 2500),
        target_role=target_role,
    )

    # ─── LLM Call ─────────────────────────────────────────────
    raw = await call_groq(system_prompt, user_prompt, temperature=0.4, max_tokens=4096)
    parsed = parse_json(raw, fallback={})
    result = normalize_rewritten_resume(parsed)

    print(f"   ✅ Resume rewritten for: {target_role}")
    return result
